web/api/v1/langserver.go

```diff
@@ -0,0 +1,134 @@
+type prometheusLangServerClient struct {
+	prometheus.Client
+	externalURL string
+	webkit      *webkit
+}
+
+func newPrometheusLangServerClient(externalURL string, webkit *webkit) prometheus.Client {
+	return &prometheusLangServerClient{
+		externalURL: externalURL,
+		webkit:      webkit,
+	}
+}
+
+// Metadata returns the first occurrence of metadata about metrics currently scraped by the metric name.
+func (c *prometheusLangServerClient) Metadata(ctx context.Context, metric string) (v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, metric, 1)
+	if len(metadata) == 0 {
+		return v1.Metadata{}, nil
+	}
+	return v1.Metadata{
+		Type: v1.MetricType(metadata[metric][0].Type),
+		Help: metadata[metric][0].Help,
+		Unit: metadata[metric][0].Unit,
+	}, nil
+}
+
+// AllMetadata returns metadata about metrics currently scraped for all existing metrics.
+func (c *prometheusLangServerClient) AllMetadata(ctx context.Context) (map[string][]v1.Metadata, error) {
+	metadata := c.webkit.metadata(ctx, "", -1)
+	result := make(map[string][]v1.Metadata)
+	for metric, infos := range metadata {
+		result[metric] = make([]v1.Metadata, 0, len(infos))
+		for i, info := range infos {
+			result[metric][i] = v1.Metadata{
+				Type: v1.MetricType(info.Type),
+				Help: info.Help,
+				Unit: info.Unit,
+			}
+		}
+	}
+	return result, nil
+}
+
+// LabelNames returns all the unique label names present in the block in sorted order.
+// If a metric is provided, then it will return all unique label names linked to the metric during a predefined period of time
+func (c *prometheusLangServerClient) LabelNames(ctx context.Context, metricName string) ([]string, error) {
+	if len(metricName) == 0 {
+		names, _, err := c.webkit.labelNames(ctx, minTime, maxTime)
+		return names, err
+	}
+	labelNames, _, err := c.webkit.series(ctx,
+		[][]*labels.Matcher{
+			{
+				&labels.Matcher{
+					Type:  labels.MatchEqual,
+					Name:  "__name__",
+					Value: metricName,
+				},
+			},
+		}, time.Now().Add(-100*time.Hour), time.Now())
+	if err != nil {
+		return nil, err
+	}
+	var result []string
+	for _, ln := range labelNames {
+		for l := range ln {
+			result = append(result, string(l))
+		}
+	}
+	return result, nil
+}
+
+// LabelValues performs a query for the values of the given label.
+func (c *prometheusLangServerClient) LabelValues(ctx context.Context, label string) ([]model.LabelValue, error) {
+	values, _, err := c.webkit.labelValues(ctx, label, minTime, maxTime)
+	if err != nil {
+		return nil, err
+	}
+	var labelValues []model.LabelValue
+	for _, value := range values {
+		labelValues = append(labelValues, model.LabelValue(value))
+	}
+	return labelValues, nil
+}
+
+// ChangeDataSource is not used in this context because we are inside prometheus.
+// So datasource/url doesn't mean anything and cannot be changed
+func (c *prometheusLangServerClient) ChangeDataSource(_ string) error {
+	return nil
+}
+
+// GetURL is returning the url used to contact the prometheus server
```

Comment:
Why is this needed if all communication is internal?

Reply:
It's part of the interface provided by the language server for a generic data source.

For a normal language server it's used to provide links to view a query in the Prometheus Web UI.

If the language server is used to provide its REST API `GetURL` is never called.

Reply:
It's part of the interface provided by the language server for a generic data source.

For a normal language server it's used to provide links to view a query in the Prometheus Web UI.

If the language server is used to provide its REST API `GetURL` is never called.

Reply:
Could we provide a dummy value then rather than plumbing through something that isn't used?

Reply:
Yes.

Reply:
What might be a cleaner solution is to change the interface function in the language server to allow returning errors and return an `not implemented` error here.

Reply:
we could just return an empty string. It will be simpler.

Reply:
That'd be ok, too.

If we end up making changes to this interface because of other reasons, we can still add the error return value then.

Reply:
usage of the externalURL has been removed. The method returns an empty string